pkg/blockbuilder/controller.go

```diff
@@ -143,11 +152,10 @@ func (l *PartitionJobController) Process(ctx context.Context, offsets Offsets, c
 
 		converted := make([]AppendInput, 0, len(records))
 		for _, record := range records {
-			offset := records[len(records)-1].Offset
-			if offset >= offsets.Max {
-				break
+			if record.Offset >= offsets.Max {
+				return lastOffset, nil
```

Comment:
I think this needs a break instead so we can still send partial results through the channel. Otherwise, we'll return before sending any `AppendInputs` to the channel. This can cause problems if half the batch we're looking at is in the offset bounds we want and half is greater.